### PR TITLE
r/aws_licensemanager: Add aws_licensemanager_association resource

### LINKED ISSUE #6654: Feature Request: License Manager
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### New or Affected Resource(s)

* `aws_licensemanager_license_configuration` (New resource)
* `aws_licensemanager_settings` (New resource)
* `aws_launch_template` (Affected resource)

### Potential Terraform Configuration

```hcl
resource "aws_licensemanager_license_configuration" "example" {
  name                     = "Example"
  description              = "Example"
  license_count            = 10
  license_count_hard_limit = true
  license_counting_type    = "Socket"

  license_rules = [
    "#minimumSockets=3"
  ]

  tags {
    foo = "barr"
  }
}
```

### References

* https://docs.aws.amazon.com/license-manager/latest/APIReference/API_CreateLicenseConfiguration.html
* https://docs.aws.amazon.com/license-manager/latest/APIReference/API_GetLicenseConfiguration.html
* https://docs.aws.amazon.com/license-manager/latest/APIReference/API_UpdateLicenseConfiguration.html
* https://docs.aws.amazon.com/license-manager/latest/APIReference/API_DeleteLicenseConfiguration.html

### PR DESCRIPTION
**Pull request by gazoakley**
Closes #6654

Changes proposed in this pull request:

* **New Resource:** `aws_licensemanager_association`
* resource/aws_launch_template: Add `license_specification` argument

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSLicenseManagerAssociation_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSLicenseManagerAssociation_basic -timeout 120m
=== RUN   TestAccAWSLicenseManagerAssociation_basic
=== PAUSE TestAccAWSLicenseManagerAssociation_basic
=== CONT  TestAccAWSLicenseManagerAssociation_basic
--- PASS: TestAccAWSLicenseManagerAssociation_basic (140.06s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	140.116s

$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSLaunchTemplate_licenseSpecification'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSLaunchTemplate_licenseSpecification -timeout 120m
=== RUN   TestAccAWSLaunchTemplate_licenseSpecification
=== PAUSE TestAccAWSLaunchTemplate_licenseSpecification
=== CONT  TestAccAWSLaunchTemplate_licenseSpecification
--- PASS: TestAccAWSLaunchTemplate_licenseSpecification (45.25s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	45.293s
```

### CODE REVIEW

File: aws/resource_aws_launch_template.go [hunk 1]
```diff
@@ -308,6 +308,20 @@ func resourceAwsLaunchTemplate() *schema.Resource {
 				Optional: true,
 			},
 
+			"license_specification": {
+				Type:     schema.TypeList,
```
Comment (bflad) [inline aws/resource_aws_launch_template.go]: Two questions:
* Does the API currently allow more than 1 specification? If not, we can set `MaxItems: 1` and the next question is moot
* Is the ordering from the API guaranteed/matter? If not, we should use `Type: schema.TypeSet`

File: aws/resource_aws_licensemanager_association.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+		}
+
+		opts.NextToken = resp.NextToken
+	}
+}
+
+func parseLicenseManagerAssociationId(id string) (string, string, error) {
+	parts := strings.SplitN(id, ",", 2)
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("Could not parse License Manager association ID: %s", id)
```
Comment (bflad) [inline aws/resource_aws_licensemanager_association.go]: Nit: Might be helpful to show the expected ID format for import:

```suggestion
		return "", "", fmt.Errorf("Expected License Manager Association ID in format RESOURCEARN,LICENSEARN, received: %s", id)
```

File: website/docs/r/licensemanager_association.markdown [hunk 3]
```diff
@@ -0,0 +1,48 @@
+---
+layout: "aws"
+page_title: "AWS: aws_licensemanager_association"
+sidebar_current: "docs-aws-resource-licensemanager-license-configuration"
```
Comment (bflad) [inline website/docs/r/licensemanager_association.markdown]: 🍝 Copy-pasta typo and missing sidebar update in `website/aws.erb`

```suggestion
sidebar_current: "docs-aws-resource-licensemanager-association"
```

File: website/docs/r/licensemanager_association.markdown [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+```hcl
+resource "aws_licensemanager_license_configuration" "example" {
+  name                  = "Example"
+  license_counting_type = "Instance"
+}
+
+resource "aws_licensemanager_association" "example" {
+  license_configuration_arn = "${aws_licensemanager_license_configuration.example.arn}"
+  resource_arn              = ...
```
Comment (bflad) [inline website/docs/r/licensemanager_association.markdown]: Nit: We should show an example reference here and/or update the `resource_arn` argument documentation below to include types of references supported

```suggestion
  resource_arn              = "${aws_instance.example.arn}"
```

File: website/docs/r/licensemanager_association.markdown (hunk 3 above)
Comment (gazoakley) [reply website/docs/r/licensemanager_association.markdown]: Doh 🍩

**Comment by gazoakley** [comment]
Hi @bflad - updated to address review comments 👍

**Comment by bflad** [comment]
This has been released in [version 1.54.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1540-december-21-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**